### ISSUE TITLE
semantic checker node flags doesn't deal with nested classes correctly

### ISSUE THREAD

**Issue by camc314** [oxc-project/oxc#12640]
Unrelated to this PR, but co-locating the enter and exit code has revealed a bug. It's not dealing correctly with nested classes.

```js
class Outer {
  // current_node_flags.contains(NodeFlags::Class) == true
  inner = class Inner {};
  // We are still inside `class Outer` but
  // current_node_flags.contains(NodeFlags::Class) == false
}
```

_Originally posted by @overlookmotel in https://github.com/oxc-project/oxc/pull/12629#discussion_r2242928715_
